EWASM interface tests

lrettig opened the pull request:
WIP. All finished and working except `log` test. Still need to test with metering on.

CF https://github.com/ethereum/tests/pull/394.

lrettig commented:
@axic Confirmed that all tests work with metering on (via https://github.com/ethereum/tests/pull/397) except one test (`getExternalCodeSize`) which has a hardcoded code size parameter.

lrettig commented:
Squashed all my commits into a single clean commit, back to you @jwasinger

jwasinger commented on src/GeneralStateTestsFiller/stEWASMTests/utils/callCodeFiller.yml:
```diff
@@ -0,0 +1,92 @@
+          (func $main
+            (i32.store (i32.const 96) (call $callCode (i64.const 100000) (i32.const 0) (i32.const 32) (i32.const 32) (i32.const 1) (i32.const 64) (i32.const 32)))
+
+            ;; Expect that the call did not trap i.e. result was 1
```
Trapping here does nothing in terms of affecting the state and changing the test outcome.  Maybe add a store statement after this and add an expect case for `abcdabcdabcdabcdabcdabcdabcdabcdabcdabcd` to have storage after trap statement be set.

lrettig replied:
"trap" is the wrong word here. Basically I think we should check that `callCode` had a result of 1 to indicate success. I think your comment here refers to the conditional that follows and the `(unreachable)`. I got this syntax from the `ewasm-kernel` tests e.g. https://github.com/ewasm/ewasm-kernel/blob/master/tests/interface/call.wast. Is there a reason those tests include this syntax but it's not useful here?

jwasinger commented on src/GeneralStateTestsFiller/stEWASMTests/utils/callDataCopyFiller.yml:
```diff
@@ -0,0 +1,67 @@
+        (module
+          (import "ethereum" "callDataCopy" (func $callDataCopy (param i32 i32 i32)))
+          (import "ethereum" "storageStore" (func $storageStore (param i32 i32)))
+          (import "debug" "printMemHex" (func $printMemHex (param i32 i32)))
```
IMO let's try and move debug statements out of tests after we have ensured that the WAST is written properly. The code is easier to read and understand when there aren't debug statements in it.

lrettig replied:
Agreed, good catch, will fix

lrettig committed:
Add tests for a bunch of EEI interface methods

Adds tests for the following methods: callDataCopy, getCallValue,
getBlockCoinbase, revert, callCode, getExternalCodeSize,
getReturnDataSize, returnDataCopy, storageStore.

Not every test has a separate, explicit test case: storageStore,
getReturnDataSize, returnDataCopy are tested inside other test fillers.

lrettig commented:
@jwasinger I fixed these issues and rebased. No more `unreachable` (I left one since the following code relies on an invariant, like an assert), instead everything is being tested in the state tests now. Also cleaned up some leftover debugging and added some locals to make the complex tests more readable. Back to you.

jwasinger commented on src/GeneralStateTestsFiller/stEWASMTests/utils/revertFiller.yml:
```diff
@@ -0,0 +1,159 @@
+            ;; Store some data to return
+            (i32.store (i32.const 0) (i32.const 0xabcd))
+            ;; Revert and return it
+            (call $revert
```
Could you add a state changing operation before the `revert` statement?  And the expect should check that the state change was reverted.

jwasinger replied:
Also add some state changing operation afterwards, and then add another expect condition that ensures that the second state changing operation was not executed.

lrettig replied:
Done

lrettig committed:
Test a few more corner cases in revertFiller

1. Make sure that a state change before `revert` is being reverted
2. Make sure code that comes after `revert` is not called
(https://github.com/ewasm/hera/issues/121)
3. Make sure that gas charged before `revert` is being charged
(https://github.com/ewasm/hera/issues/120)

lrettig committed:
Adds createFromTransaction test filler

Tests creating a contract from transaction data.
Modify convert_wast.py to allow empty 'pre' data.

Not currently working, may be due to
https://github.com/ewasm/hera/issues/122.

lrettig committed:
Remove useGas, update expect value

Removing gas checks from this test per @jwasinger, as they will go in a
separate test. Fix formatting slightly. Also update the expected return
value from the reverted call, which is now 0x02 per the design doc.

lrettig commented:
@jwasinger I fixed the test to expect the correct return value from the reverted `call`, and removed the gas checks as we discussed. Back to you again.

Let me know if you want me to squash.
